var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -269,6 +270,104 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
 
     build_directory = 'spack-build'
 
+    @property
+    def executables(self):
+        names = [r'gcc', r'[^\w]?g\+\+', r'gfortran']
+        suffixes = [r'', r'-mp-\d+\.\d', r'-\d+\.\d', r'-\d+', r'\d\d']
+        return [r''.join(x) for x in itertools.product(names, suffixes)]
+
+    @classmethod
+    def filter_detected_exes(cls, prefix, exes_in_prefix):
+        result = []
+        for exe in exes_in_prefix:
+            # clang++ matches g++ -> clan[g++]
+            if any(x in exe for x in ('clang', 'ranlib')):
+                continue
+            # Filter out links in favor of real executables
+            if os.path.islink(exe):
+                continue
+            result.append(exe)
+        return result
+
+    @classmethod
+    def determine_version(cls, exe):
+        version_regex = re.compile(r'([\d\.]+)')
+        for vargs in ('-dumpfullversion', '-dumpversion'):
+            try:
+                output = spack.compiler.get_compiler_version_output(exe, vargs)
+                match = version_regex.search(output)
+                if match:
+                    return match.group(1)
+            except spack.util.executable.ProcessError:
+                pass
+            except Exception as e:
+                tty.debug(e)
+
+        return None
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
+        languages, compilers = set(), {}
+        for exe in exes:
+            if 'gcc' in exe:
+                languages.add('c')
+                compilers['c'] = exe
+            if 'g++' in exe:
```

Comment:
Is it possible for an executable to be named `gcc-5.3.0-g++`? Also, should these be elifs?

Reply:
I didn't encounter cases like that so far. #17525 builds on this PR and moves all the detection logic of compilers to packages. It also adds the possibility to write a `detection_tests.yaml` file (alongside `package.py`) that will be used to create mock tree structures with fake executables and to check that the correct spec is detected. If that's fine with you we can work the fine logic for compiler detection there.

Reply:
Improved a bit in 450e891